### ISSUE TITLE
 Print to browser doesn't work on Firefox snap package

### ISSUE THREAD

**Issue by ihavenomouth** [zim-desktop-wiki/zim-desktop-wiki#2025]
I'm not sure if this can be considered a bug or not because snap is the culprit. Since Ubuntu 22.04 is adopting Fierfox snap packages and snaps can't access a file outside the $HOME directory, the print to browser (Ctrl+P) is almost useless. As far as I know a snap package can't access /tmp (it uses its own tmp folder), so it doesn't open the output file automatically.
The print to browser feature is still working properly, but is not open by Firefox.

It shows an error similar to _Firefox can't find the file in /tmp/zim-17ncxc3o/print-to-browser.html_

I've tried to do this, but it didn't work either:
> mkdir /home/javier/tmp
> sudo mount --bind /tmp /home/javier/tmp/

I've changed the code in `/usr/lib/python3/dist-packages/zim/plugins/printtobrowser.py` so it use a file in my home folder instead a tmp file. And it works (I know almost nothing about python, basically I've used the open function to write the content). So it seems possible to solve the issue if we could choose the output file in a configure menu.

TL;DR: snap packages can't read files in /tmp, so print to browser doesn't work automatically.

**Comment by introt** [comment]
Agreed, not a Zim bug & a plugin preference sounds like a good workaround.

**Comment by Nick222** [comment]
I changed in Zim options default browser to Chromium - but Zim try to open print-to-browser-TMP-file in Firefox (default system browser).
Xubuntu 22.10

**Comment by bruvellu** [comment]
Print to browser also doesn't work on the [Flatpak version of Zim 0.75.2](https://flathub.org/apps/org.zim_wiki.Zim). I guess it's a related issue to the snap.

The file is created fine in Zim's flatpak tmp directory (example `~/.var/app/org.zim_wiki.Zim/cache/tmp/zim-45sx6bg1/print-to-browser.html`), but it doesn't open in the browser.

Since it's not clear to me where the solution is (Zim code or [Flatpak packaging](https://github.com/flathub/org.zim_wiki.Zim)), I'm reporting it here for now.

**Comment by vmagnin** [comment]
Same problem for me since Firefox is in the Ubuntu Snap system. The workaround is to open the file manually in Firefox, looking in the `\tmp` directory for the subdirectory created by Zim. It works but it is a regression.

For example, Firefox will not find that file:
```txt
file:///tmp/zim-ltq_ncve/print-to-browser.html
```
but when you open it manually, it will display that URI:
```txt
file:///run/user/1000/doc/61734754/print-to-browser.html
```

**Comment by Nick222** [comment]
I find this file - print-to-browser.html - but cannot open it by Firefox.

It opened only after manually copying in my home directory.

Xubuntu 24.10
Zim 0.76

**Comment by vmagnin** [comment]
@Nick222 
I can access it with Kubuntu 24.10.

Maybe your Firefox has not rights to write in `/tmp/` ?

In the case of Thunderbird, I had problems with `/tmp/` and USB devices, fixed with those commands:

```bash
$ snap connections thunderbird
$ snap connect thunderbird:removable-media
```
The first one is showing all the "connections", the second one is adding one.

**Comment by FredworkLemmas** [comment]
I'd been having trouble with this as well, I tracked down how this plugin works, and found it depends on the python webbrowser library which will use the BROWSER environment variable, if it's defined.  I set it in .bashrc (and logged out/in) and it's working as it should now.
